File: BinarySearch.py
class BinarySearch:
    def binary_search_recursive(self, arr, target, start, end):  # 재귀 구현
        mid = (start + end) // 2
        if arr[mid] == target:
            return mid

        if start > end:
            return -1

        if arr[mid] > target:
            return self.binary_search_recursive(arr, target, start, mid - 1)
        else:
            return self.binary_search_recursive(arr, target, mid + 1, end)

    def binary_search_loop(self, arr, target, start, end):  # 반복문 구현
        while start <= end:
            mid = (start + end) // 2
            if arr[mid] == target:
                return mid
            if arr[mid] > target:
                end = mid - 1
            else:
                start = mid + 1
        return -1

File: Go/search/binary_search.go
package search

import (
	"golang.org/x/exp/constraints"
)

func BinarySearchRecursive[T constraints.Ordered](arr []T, left int, right int, value T) (index int, found bool) {
	if left <= right {
		mid := (left + right) / 2
		if arr[mid] == value {
			return mid, true
		} else if arr[mid] < value {
			return BinarySearchRecursive[T](arr, mid+1, right, value)
		} else {
			return BinarySearchRecursive[T](arr, left, mid-1, value)
		}
	}
	return index, found
}

func BinarySearchRepeated[T constraints.Ordered](arr []T, value T) (index int, found bool) {
	left, right := 0, len(arr)-1
	for left <= right {
		mid := (left + right) / 2
		if arr[mid] == value {
			return mid, true
		} else if arr[mid] < value {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return index, found
}
